src/sync/mod.rs

```diff
@@ -0,0 +1,146 @@
+use crate::Result;
+use std::collections::VecDeque;
+use crate::{Error, wpactrl::ClientInternal, ClientBuilder};
+use futures::executor::block_on;
+/// A connection to `wpa_supplicant` / `hostapd`
+pub struct Client(ClientInternal);
+
+impl Client {
+    /// Creates a builder for a `wpa_supplicant` / `hostapd` connection
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let wpa = wpactrl::Client::builder().open().unwrap();
+    /// ```
+    #[must_use]
+    pub fn builder() -> ClientBuilder {
+        ClientBuilder::default()
+    }
+
+    /// Register as an event monitor for control interface messages
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut wpa = wpactrl::Client::builder().open().unwrap();
+    /// let wpa_attached = wpa.attach().unwrap();
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// * [`Error::Attach`] - Unexpected (non-OK) response
+    /// * [`Error::Io`] - Low-level I/O error
+    /// * [`Error::Utf8ToStr`] - Corrupted message or message with non-UTF8 characters
+    /// * [`Error::Wait`] - Failed to wait on underlying Unix socket
+    pub fn attach(mut self) -> Result<ClientAttached> {
+        // FIXME: None closure would be better
+        if block_on(self.0.request("ATTACH", |_: &str| ()))? == "OK\n" {
+            Ok(ClientAttached(self.0, VecDeque::new()))
+        } else {
+            Err(Error::Attach)
+        }
+    }
+
+    /// Send a command to `wpa_supplicant` / `hostapd`.
+    ///
+    /// Commands are generally identical to those used in `wpa_cli`,
+    /// except all uppercase (eg `LIST_NETWORKS`, `SCAN`, etc)
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut wpa = wpactrl::Client::builder().open().unwrap();
+    /// assert_eq!(wpa.request("PING").unwrap(), "PONG\n");
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// * [`Error::Io`] - Low-level I/O error
+    /// * [`Error::Utf8ToStr`] - Corrupted message or message with non-UTF8 characters
+    /// * [`Error::Wait`] - Failed to wait on underlying Unix socket
+    pub fn request(&mut self, cmd: &str) -> Result<String> {
```

Comment:
I think `cb` should be passed through. Although, IIRC, this may have just been to mirror the C API.